container_api/src/interface.rs

```diff
@@ -17,133 +17,185 @@ pub trait DispatchRpc {
     fn handler(self) -> IoHandler;
 }
 
-/// ContainerApiDispatcher exposes some subset of the Container API,
-/// including zome function calls as well as admin functionality.
-/// Each interface has their own dispatcher, and each may be configured differently.
-pub struct ContainerApiDispatcher {
-    instances: InstanceMap,
-    pub io: IoHandler,
-}
-
-/// Implements routing for JSON-RPC calls:
+/// ContainerApiBuilder creates IoHandlers that implement RPCs for exposure
+/// through interfaces or bridges.
+/// This includes zome function calls as well as admin functionality.
+///
+/// Examples for method names are:
 /// {instance_id}/{zome}/{cap}/{func} -> a zome call
 /// info/list_instances               -> Map of InstanceConfigs, keyed by ID
 /// admin/...                         -> TODO
-impl ContainerApiDispatcher {
-    pub fn new(config: &Configuration, instances: InstanceMap) -> Self {
-        let instance_configs = config
+///
+/// Each interface has their own handler, and each may be configured differently.
+/// This builder makes it convenient to create handlers with different configurations.
+///
+/// Call any sequence of with_* functions on a ContainerApiBuilder object and finalize
+/// with spawn() to retrieve the IoHandler.
+pub struct ContainerApiBuilder {
+    instances: InstanceMap,
+    instance_configs: HashMap<String, InstanceConfiguration>,
+    io: Box<IoHandler>,
+}
+
+impl ContainerApiBuilder {
+    pub fn new() -> Self {
+        ContainerApiBuilder {
+            instances: HashMap::new(),
+            instance_configs: HashMap::new(),
+            io: Box::new(IoHandler::new()),
+        }
+    }
+
+    /// Finish the building and retrieve the populated handler
+    pub fn spawn(mut self) -> IoHandler {
+        self.setup_info_api();
+        *self.io
+    }
+
+    /// Adds a "info/instances" method that returns a JSON object describing all registered
+    /// instances we have a config for.
+    fn setup_info_api(&mut self) {
+        let instance_configs = self.instance_configs.clone();
+
+        let configs: Vec<_> = self
             .instances
             .iter()
-            .map(|inst| (inst.id.clone(), inst.clone()))
+            .filter(|&(name, _)| instance_configs.contains_key(name))
```

Comment:
The `info/instances` is how the UI knows what DNA hashes it can refer to, pretty essential if the UI hard-codes DNA hashes as the fundamental address of a DNA, which I think it should. So if someone writes a container that doesn't pass an config for every instance, it will break the UI.